include/ruby/internal/core/robject.h

```diff
@@ -72,24 +70,57 @@ ROBJECT_NUMIV(VALUE obj)
         return ROBJECT_EMBED_LEN_MAX;
     }
     else {
-        return ROBJECT(obj)->as.heap.numiv;
+        return (uint32_t)NUM2INT(ROBJECT(obj)->as.heap.ivptr[0]);
+    }
+}
+
+RBIMPL_ATTR_PURE_UNLESS_DEBUG()
+RBIMPL_ATTR_ARTIFICIAL()
+static inline void
+ROBJECT_NUMIV_SET(VALUE obj, uint32_t numiv)
+{
+    RBIMPL_ASSERT_TYPE(obj, RUBY_T_OBJECT);
+
+    if (!RB_FL_ANY_RAW(obj, ROBJECT_EMBED)) {
+        ROBJECT(obj)->as.heap.ivptr[0] = INT2NUM(numiv);
     }
 }
 
 RBIMPL_ATTR_PURE_UNLESS_DEBUG()
 RBIMPL_ATTR_ARTIFICIAL()
-static inline VALUE *
-ROBJECT_IVPTR(VALUE obj)
+static inline VALUE
+ROBJECT_IVPTR(VALUE obj, uint32_t idx)
```

Comment:
This doesn't seem **PTR** now.

Reply:
If the performance seems good I will fix the function names. This is just an experiment for now 😅